src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Cryptography;
+
+namespace Internal.Cryptography
+{
+    //
+    // Common infrastructure for crypto symmetric algorithms that use Cng.
+    //
+    internal struct CngSymmetricAlgorithmCore
+    {
+        /// <summary>
+        /// Configures the core to use plaintext keys (to be auto-generated when first needed.)
+        /// </summary>
+        public CngSymmetricAlgorithmCore(string algorithm, ICngSymmetricAlgorithm outer)
+        {
+            _algorithm = algorithm;
+            _outer = outer;
+
+            _keyName = null; // Setting _keyName to null signifies that this object is based on a plaintext key, not a stored CNG key.
+            _provider = null;
+            _optionOptions = CngKeyOpenOptions.None;
+        }
+
+        /// <summary>
+        /// Constructs the core to use a stored CNG key. 
+        /// </summary>
+        public CngSymmetricAlgorithmCore(string algorithm, ICngSymmetricAlgorithm outer, string keyName, CngProvider provider, CngKeyOpenOptions openOptions)
+        {
+            if (keyName == null)
+                throw new ArgumentNullException("keyName");
+            if (provider == null)
+                throw new ArgumentNullException("provider");
+
+            _algorithm = algorithm;
+            _outer = outer;
+
+            _keyName = keyName;
+            _provider = provider;
+            _optionOptions = openOptions;
+
+            using (CngKey cngKey = ProduceCngKey())
+            {
+                CngAlgorithm actualAlgorithm = cngKey.Algorithm;
+                if (algorithm != actualAlgorithm.Algorithm)
+                    throw new CryptographicException(SR.Format(SR.Cryptography_CngKeyWrongAlgorithm, actualAlgorithm.Algorithm, algorithm));
+
+                _outer.BaseKeySize = cngKey.KeySize;
+            }
+        }
+
+        /// <summary>
+        /// Note! This can and likely will throw if the algorithm was given a hardware-based key.
+        /// </summary>
+        public byte[] GetKeyIfExportable()
+        {
+            if (KeyInPlainText)
+            {
+                return _outer.BaseKey;
+            }
+            else
+            {
+                using (CngKey cngKey = ProduceCngKey())
+                {
+                    return cngKey.GetSymmetricKeyDataIfExportable(_algorithm);
+                }
+            }
+        }
+
+        public void SetKey(byte[] key)
+        {
+            _outer.BaseKey = key;
+            _keyName = null; // Setting _keyName to null signifies that this object is now based on a plaintext key, not a stored CNG key.
```

Comment:
Fair question. I’m no fan of the way these Crypto classes let you reassign everything under the sun instead of specifying things once and for all in a constructor.

On the other hand, that is the pattern, so I didn’t want to change it for this one case.